File: test/data/local/beverage/beverage_store_test.dart
import 'package:drift/drift.dart' show Value;
import 'package:drift/native.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:openvitals/data/local/beverage/beverage_entity.dart';
import 'package:openvitals/data/local/beverage/beverage_store.dart';
import 'package:openvitals/data/local/open_vitals_database.dart';
import 'package:openvitals/data/prefs/preferences_repository.dart';
import 'package:openvitals/domain/model/caffeine_models.dart';
import 'package:openvitals/domain/model/nutrition_models.dart';
import 'package:shared_preferences/shared_preferences.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  late OpenVitalsDatabase db;
  late PreferencesRepository prefsRepo;
  late BeverageStore store;

  setUp(() async {
    SharedPreferences.setMockInitialValues({});
    final prefs = await SharedPreferences.getInstance();
    prefsRepo = PreferencesRepository(prefs);
    db = OpenVitalsDatabase(NativeDatabase.memory());
    store = BeverageStore(db.beverageDao, prefsRepo);
  });

  tearDown(() async {
    await db.close();
  });

  test('beverages() seeds preloaded defaults on first access', () async {
    final beverages = await store.beverages();

    expect(beverages.length, BeverageEntity.preloadedDefaults().length);
    expect(beverages[0].id, 'openvitals-still-water');
    expect(beverages[1].id, 'openvitals-gasified-water');
    // The migration flag is flipped once the store initializes.
    expect(prefsRepo.hasMigratedHydrationBeveragesToRoom(), isTrue);
  });

  test('save() inserts a new active drink with the next sort order', () async {
    await store.beverages(); // force seeding
    const drink = CustomHydrationDrink(
      id: 'my-smoothie',
      name: 'Smoothie',
      volumeMilliliters: 350.0,
      hydrationMultiplier: 0.8,
      nutrientValues: {NutritionNutrient.energy: 180.0},
      category: CaffeineSourceCategory.other,
    );

    await store.save(drink);
    final beverages = await store.beverages();

    final saved = beverages.firstWhere((d) => d.id == 'my-smoothie');
    expect(saved.name, 'Smoothie');
    expect(saved.volumeMilliliters, closeTo(350.0, 0.001));
    expect(saved.hydrationMultiplier, closeTo(0.8, 0.001));
    expect(saved.category, CaffeineSourceCategory.other);
    expect(saved.nutrientValues[NutritionNutrient.energy], closeTo(180.0, 0.001));
  });

  test('delete() soft-deletes and hides the drink from active listing',
      () async {
    await store.beverages();
    await store.delete('openvitals-still-water');

    final beverages = await store.beverages();
    expect(
      beverages.where((d) => d.id == 'openvitals-still-water'),
      isEmpty,
    );
    // Soft delete keeps the row, just flips is_deleted.
    final row = await db.beverageDao.beverageById('openvitals-still-water');
    expect(row, isNotNull);
    expect(row!.isDeleted, isTrue);
  });

  test('moveToCategory() updates the persisted category', () async {
    await store.beverages();
    await store.moveToCategory('openvitals-still-water', CaffeineSourceCategory.tea);

    final row = await db.beverageDao.beverageById('openvitals-still-water');
    expect(row!.category, CaffeineSourceCategory.tea.storageName);
  });

  test('reorder() reindexes provided ids first, keeping the rest after',
      () async {
    await store.beverages();
    await store.reorder(['openvitals-gasified-water', 'openvitals-still-water']);

    final beverages = await store.beverages();
    expect(beverages[0].id, 'openvitals-gasified-water');
    expect(beverages[1].id, 'openvitals-still-water');
  });

  test('DAO nextSortOrder returns max+1', () async {
    await db.beverageDao.insertDefaults([
      const BeverageEntity(
        id: 'a',
        name: 'A',
        category: null,
        volumeMilliliters: 100,
        hydrationMultiplier: 1,
        isPreloaded: false,
        sortOrder: 4,
      ),
    ]);
    expect(await db.beverageDao.nextSortOrder(), 5);
  });

  test('DAO insertDefaults ignores conflicts on existing ids', () async {
    const entity = BeverageEntity(
      id: 'dup',
      name: 'First',
      category: null,
      volumeMilliliters: 100,
      hydrationMultiplier: 1,
      isPreloaded: false,
      sortOrder: 0,
    );
    await db.beverageDao.insertDefaults([entity]);
    await db.beverageDao.insertDefaults([
      entity.copyWith(name: 'Second'),
    ]);
    final row = await db.beverageDao.beverageById('dup');
    expect(row!.name, 'First');
  });

  test('DAO upsert replaces the row and can clear the delete flag', () async {
    await db.beverageDao.upsert(
      const BeverageEntity(
        id: 'x',
        name: 'X',
        category: null,
        volumeMilliliters: 100,
        hydrationMultiplier: 1,
        isPreloaded: false,
        isDeleted: true,
        sortOrder: 0,
      ),
    );
    await (db.update(db.beverages)..where((b) => b.id.equals('x')))
        .write(const BeveragesCompanion(name: Value('X2')));
    final row = await db.beverageDao.beverageById('x');
    expect(row!.name, 'X2');
  });
}

File: app/src/test/kotlin/tech/mmarca/openvitals/data/local/beverage/BeverageStoreTest.kt
package tech.mmarca.openvitals.data.local.beverage

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkStatic
import io.mockk.unmockkStatic
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import tech.mmarca.openvitals.data.repository.PreferencesRepository
import tech.mmarca.openvitals.devices.FakeSharedPreferences
import tech.mmarca.openvitals.domain.model.BeverageCategory
import tech.mmarca.openvitals.domain.model.CustomHydrationDrink
import tech.mmarca.openvitals.domain.model.NutritionNutrient

/**
 * In-memory stand-in for [BeverageDao]. There is no SQLite in the unit suite, so each
 * `@Query`'s SQL is reimplemented over a map. The `@Transaction` default methods run for real.
 */
private class FakeBeverageDao : BeverageDao {
    val rows = linkedMapOf<String, BeverageEntity>()

    override suspend fun activeBeverages(): List<BeverageEntity> =
        rows.values
            .filterNot { it.isDeleted }
            .sortedWith(compareBy({ it.sortOrder }, { it.name.lowercase() }))

    override suspend fun beverageById(id: String): BeverageEntity? = rows[id]

    override suspend fun nextSortOrder(): Int =
        (rows.values.maxOfOrNull { it.sortOrder } ?: -1) + 1

    override suspend fun insertDefaults(beverages: List<BeverageEntity>) {
        // OnConflictStrategy.IGNORE: an existing id keeps its stored row.
        beverages.forEach { entity -> rows.putIfAbsent(entity.id, entity) }
    }

    override suspend fun upsert(beverage: BeverageEntity) {
        rows[beverage.id] = beverage
    }

    override suspend fun softDelete(id: String) {
        rows[id]?.let { rows[id] = it.copy(isDeleted = true) }
    }

    override suspend fun updateCategory(id: String, category: String?) {
        rows[id]?.let { rows[id] = it.copy(category = category) }
    }

    override suspend fun updateSortOrder(id: String, sortOrder: Int) {
        rows[id]?.let { rows[id] = it.copy(sortOrder = sortOrder) }
    }

    override suspend fun deleteAll() {
        rows.clear()
    }

    override suspend fun insertAll(beverages: List<BeverageEntity>) {
        beverages.forEach { rows[it.id] = it }
    }
}

/** The store over [FakeBeverageDao] and a [FakeSharedPreferences]-backed [PreferencesRepository]. */
class BeverageStoreTest {

    private lateinit var dao: FakeBeverageDao
    private lateinit var prefs: FakeSharedPreferences
    private lateinit var preferencesRepository: PreferencesRepository
    private lateinit var store: BeverageStore

    @Before
    fun setUp() {
        mockkStatic(Log::class)
        every { Log.w(any(), any<String>()) } returns 0
        every { Log.e(any(), any<String>()) } returns 0
        dao = FakeBeverageDao()
        prefs = FakeSharedPreferences()
        val context = mockk<Context> {
            every {
                getSharedPreferences(PreferencesRepository.PREFS_FILE, Context.MODE_PRIVATE)
            } returns prefs as SharedPreferences
        }
        preferencesRepository = PreferencesRepository(context)
        store = BeverageStore(dao, preferencesRepository)
    }

    @After
    fun tearDown() {
        unmockkStatic(Log::class)
    }

    @Test
    fun `beverages seeds preloaded defaults on first access`() {
        val beverages = store.beverages()

        assertEquals(BeverageEntity.preloadedDefaults().size, beverages.size)
        assertEquals("openvitals-still-water", beverages[0].id)
        assertEquals("openvitals-gasified-water", beverages[1].id)
        // The migration flag is flipped once the store initializes.
        assertTrue(preferencesRepository.hasMigratedHydrationBeveragesToRoom())
    }

    @Test
    fun `save inserts a new active drink with the next sort order`() {
        store.beverages() // force seeding
        val drink = CustomHydrationDrink(
            id = "my-smoothie",
            name = "Smoothie",
            volumeMilliliters = 350.0,
            hydrationMultiplier = 0.8,
            nutrientValues = mapOf(NutritionNutrient.ENERGY to 180.0),
            category = BeverageCategory.OTHER,
        )

        store.save(drink)
        val beverages = store.beverages()

        val saved = beverages.first { it.id == "my-smoothie" }
        assertEquals("Smoothie", saved.name)
        assertEquals(350.0, saved.volumeMilliliters, 0.001)
        assertEquals(0.8, saved.hydrationMultiplier, 0.001)
        assertEquals(BeverageCategory.OTHER, saved.category)
        assertEquals(180.0, saved.nutrientValues.getValue(NutritionNutrient.ENERGY), 0.001)
        // The next free sort order, i.e. after the whole preloaded catalog.
        assertEquals(BeverageEntity.preloadedDefaults().size, dao.rows.getValue("my-smoothie").sortOrder)
    }

    @Test
    fun `delete soft-deletes and hides the drink from active listing`() {
        store.beverages()
        store.delete("openvitals-still-water")

        val beverages = store.beverages()
        assertNull(beverages.firstOrNull { it.id == "openvitals-still-water" })
        // Soft delete keeps the row, just flips is_deleted.
        val row = dao.rows["openvitals-still-water"]
        assertNotNull(row)
        assertTrue(row!!.isDeleted)
    }

    @Test
    fun `moveToCategory updates the persisted category`() {
        store.beverages()
        store.moveToCategory("openvitals-still-water", BeverageCategory.TEA)

        assertEquals(BeverageCategory.TEA.name, dao.rows.getValue("openvitals-still-water").category)
    }

    @Test
    fun `reorder reindexes provided ids first, keeping the rest after`() {
        store.beverages()
        store.reorder(listOf("openvitals-gasified-water", "openvitals-still-water"))

        val beverages = store.beverages()
        assertEquals("openvitals-gasified-water", beverages[0].id)
        assertEquals("openvitals-still-water", beverages[1].id)
    }
}
